### ISSUE TITLE
How to use --keep_lcc ???

### ISSUE THREAD

**Issue by cthulhu-rises** [NVlabs/neuralangelo#144]
Unsurprisingly there is no info on where in the extraction syntax to throw this in. I always get an error regardless of where I put it.

**Comment by hypertechnic** [comment]
Same issue for me

**Comment by Bombaninha** [comment]
The parameter `keep-lcc` is provided in the script `extract_mesh.py`.

```bash
CHECKPOINT=logs/${GROUP}/${NAME}/xxx.pt
OUTPUT_MESH=xxx.ply
CONFIG=logs/${GROUP}/${NAME}/config.yaml
RESOLUTION=2048
BLOCK_RES=128
GPUS=1  # use >1 for multi-GPU mesh extraction
torchrun --nproc_per_node=${GPUS} projects/neuralangelo/scripts/extract_mesh.py \
    --config=${CONFIG} \
    --checkpoint=${CHECKPOINT} \
    --output_file=${OUTPUT_MESH} \
    --resolution=${RESOLUTION} \
    --block_res=${BLOCK_RES}
    --keep_lcc
```

According to the arg parser, this parameter has the following effect:

> Keep only largest connected component. May remove thin structures.

In the same file, there is a call to the function `extract_mesh`, declared in `projects.neuralangelo.utils.mesh`, with the following body:

**Calling extract_mesh**
```python
mesh = extract_mesh(sdf_func=sdf_func, bounds=bounds, intv=(2.0 / args.resolution), block_res=args.block_res, texture_func=texture_func, filter_lcc=args.keep_lcc)
```

**extract_mesh definition**
```python
def extract_mesh(sdf_func, bounds, intv, block_res=64, texture_func=None, filter_lcc=False):
    lattice_grid = LatticeGrid(bounds, intv=intv, block_res=block_res)
    data_loader = get_lattice_grid_loader(lattice_grid)
    mesh_blocks = []
    if is_master():
        data_loader = tqdm(data_loader, leave=False)
    for it, data in enumerate(data_loader):
        xyz = data["xyz"][0]
        xyz_cuda = xyz.cuda()
        sdf_cuda = sdf_func(xyz_cuda)[..., 0]
        sdf = sdf_cuda.cpu()
        mesh = marching_cubes(sdf.numpy(), xyz.numpy(), intv, texture_func, filter_lcc)
        mesh_blocks.append(mesh)
    mesh_blocks_gather = [None] * get_world_size()
    if dist.is_initialized():
        dist.all_gather_object(mesh_blocks_gather, mesh_blocks)
    else:
        mesh_blocks_gather = [mesh_blocks]
    if is_master():
        mesh_blocks_all = [mesh for mesh_blocks in mesh_blocks_gather for mesh in mesh_blocks
                           if mesh.vertices.shape[0] > 0]
        mesh = trimesh.util.concatenate(mesh_blocks_all)
        return mesh
    else:
        return None
```

This parameter is passed to the `marching_cubes` function, which ultimately operates the `filter_largest_cc`.

**Marching cubes**
```python
def marching_cubes(sdf, xyz, intv, texture_func, filter_lcc):
    # marching cubes
    V, F = mcubes.marching_cubes(sdf, 0.)
    if V.shape[0] > 0:
        V = V * intv + xyz[0, 0, 0]
        if texture_func is not None:
            C = texture_func(V)
            mesh = trimesh.Trimesh(V, F, vertex_colors=C)
        else:
            mesh = trimesh.Trimesh(V, F)
        mesh = filter_points_outside_bounding_sphere(mesh)
        mesh = filter_largest_cc(mesh) if filter_lcc else mesh
    else:
        mesh = trimesh.Trimesh()
    return mesh
```

```python
def filter_largest_cc(mesh):
    components = mesh.split(only_watertight=False)
    areas = np.array([c.area for c in components], dtype=float)
    if len(areas) > 0 and mesh.vertices.shape[0] > 0:
        new_mesh = components[areas.argmax()]
    else:
        new_mesh = trimesh.Trimesh()
    return new_mesh
```

From what I can observe, this function performs the following steps:

**1. Splitting the Mesh:**

The split function is used to split the mesh into its connected components. 

**2. Calculating Areas:**

For each connected component, the function calculates the area of its surface.

**3. Determining the Largest Connected Component:**

The areas of the connected components are stored in an array (areas).
If there is at least one component `(len(areas) > 0)` and the original mesh has vertices `(mesh.vertices.shape[0] > 0)`, the function identifies the largest connected component based on the maximum area `(areas.argmax())`. The corresponding component is stored in `new_mesh`.

**4. Returning the Result:**

If there are no areas or vertices, or if an issue occurred, an empty `trimesh.Trimesh()` object is returned. Otherwise, the largest connected component is returned as a new mesh.

In summary, this function filters out all connected components of a mesh except for the largest one based on their surface areas. It ensures that the returned mesh is the largest connected component of the original mesh.

I hope this trace helps you in some way.
